design/src/rfcs/rfc0025_better_constraint_violations.md

```diff
@@ -0,0 +1,511 @@
+user has to provide a map whose values are already constrained `LengthString`s
+to the `try_from` constructor, which only enforces the map's `@length` trait.
+
+The reason why these seemingly "impossible violations" are being generated is
+because they can arise during request deserialization. Indeed, the server
+framework deserializes requests into **fully unconstrained types**. These are
+types holding unconstrained types all the way through their closures. For
+instance, in the case of structure shapes, builder types (the unconstrained
+type corresponding to the structure shape) [hold
+builders][builders-of-builders-pr] all the way down.
+
+In the case of the above model, below is the alternate `pub(crate)` constructor
+the server framework uses upon deserialization. Observe how
+`LengthMapOfLengthStringsUnconstrained` is _fully unconstrained_ and how the
+`try_from` constructor can yield `ConstraintViolation::Value`.
+
+```rust
+pub(crate) mod length_map_of_length_strings_unconstrained {
+    #[derive(Debug, Clone)]
+    pub(crate) struct LengthMapOfLengthStringsUnconstrained(
+        pub(crate) std::collections::HashMap<std::string::String, std::string::String>,
+    );
+
+    impl std::convert::TryFrom<LengthMapOfLengthStringsUnconstrained>
+        for crate::model::LengthMapOfLengthStrings
+    {
+        type Error = crate::model::length_map_of_length_strings::ConstraintViolation;
+        fn try_from(value: LengthMapOfLengthStringsUnconstrained) -> Result<Self, Self::Error> {
+            let res: Result<
+                std::collections::HashMap<std::string::String, crate::model::LengthString>,
+                Self::Error,
+            > = value
+                .0
+                .into_iter()
+                .map(|(k, v)| {
+                    let v: crate::model::LengthString = k.try_into().map_err(Self::Error::Key)?;
+
+                    Ok((k, v))
+                })
+                .collect();
+            let hm = res?;
+            Self::try_from(hm)
+        }
+    }
+}
+```
+
+In conclusion, the user is currently exposed to an internal detail of how the
+framework operates that has no bearing on their application code. They
+shouldn't be exposed to impossible constraint violation variants in their Rust
+docs, nor have to `match` on these variants when handling errors.
+
+Note: [this comment] alludes to the problem described above.
+
+[this comment]: https://github.com/awslabs/smithy-rs/blob/27020be3421fb93e35692803f9a795f92feb1d19/codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/MapConstraintViolationGenerator.kt#L66-L69
+
+### Solution proposal
+
+The problem can be mitigated by adding `#[doc(hidden)]` to the internal
+variants and `#[non_exhaustive]` to the enum. We're already doing this in some
+constraint violation types.
+
+However, a "less leaky" solution is achieved by _splitting_ the constraint
+violation type into two types:
```

Comment:
From this document I cannot really tell if you are proposing the mitigation or the more complex solution of splitting the constraints into two types. I believe the latter, but it's better to be more clear I think.

Reply:
The document is proposing the "less leaky" solution, splitting the constraint violation type into two types. I'll rephrase.